apps/small_filters/process.h

```diff
@@ -0,0 +1,526 @@
+#ifndef PROCESS_H
+#define PROCESS_H
+
+#include "HalideRuntimeHexagonHost.h"
+#include "HalideBuffer.h"
+
+#define DECL_WEAK_PIPELINE3S(x) int x##_cpu(buffer_t *, buffer_t *, buffer_t *) __attribute__((weak)); \
+    int x##_hvx128(buffer_t *, buffer_t *, buffer_t *) __attribute__((weak)); \
+    int x##_hvx64(buffer_t *, buffer_t *, buffer_t *) __attribute__((weak));
+
+#define DECL_WEAK_PIPELINE2S(x) int x##_cpu(buffer_t *, buffer_t *) __attribute__((weak)); \
+    int x##_hvx128(buffer_t *, buffer_t *) __attribute__((weak)); \
+    int x##_hvx64(buffer_t *, buffer_t *) __attribute__((weak));
+
+
+#ifdef CONV3X3A16
+#include "conv3x3a16_hvx128.h"
+#include "conv3x3a16_hvx64.h"
+#include "conv3x3a16_cpu.h"
+#else
+DECL_WEAK_PIPELINE3S(conv3x3a16)
+#endif
+
+#ifdef DILATE3X3
+#include "dilate3x3_hvx128.h"
+#include "dilate3x3_hvx64.h"
+#include "dilate3x3_cpu.h"
+#else
+DECL_WEAK_PIPELINE2S(dilate3x3)
+#endif
+
+#ifdef MEDIAN3X3
+#include "median3x3_hvx128.h"
+#include "median3x3_hvx64.h"
+#include "median3x3_cpu.h"
+#else
+DECL_WEAK_PIPELINE2S(median3x3)
+#endif
+
+#ifdef GAUSSIAN5X5
+#include "gaussian5x5_hvx128.h"
+#include "gaussian5x5_hvx64.h"
+#include "gaussian5x5_cpu.h"
+#else
+DECL_WEAK_PIPELINE2S(gaussian5x5)
+#endif
+
+#ifdef SOBEL
+#include "sobel_hvx128.h"
+#include "sobel_hvx64.h"
+#include "sobel_cpu.h"
+#else
+DECL_WEAK_PIPELINE2S(sobel)
+#endif
+
+#ifdef CONV3X3A32
+#include "conv3x3a32_hvx128.h"
+#include "conv3x3a32_hvx64.h"
+#include "conv3x3a32_cpu.h"
+#else
+DECL_WEAK_PIPELINE3S(conv3x3a32)
+#endif
+
+
+enum bmark_run_mode_t {
+    hvx64 = 1,
+    hvx128 = 2,
+    cpu = 3
+};
+
+template <typename T>
+T clamp(T val, T min, T max) {
+    if (val < min)
+        return min;
+    if (val > max)
+        return max;
+    return val;
+}
+
+typedef int (*pipeline2)(buffer_t *, buffer_t *);
+typedef int (*pipeline3)(buffer_t *, buffer_t *, buffer_t *);
+
+struct PipelineDescriptorBase {
+    virtual void init() = 0;
+    virtual const char * name() = 0;
+    virtual int run(bmark_run_mode_t mode) = 0;
+    virtual bool verify(int W, int H) = 0;
+    virtual bool defined() = 0;
+};
+
+template <typename T1>
+struct PipelineDescriptor : public PipelineDescriptorBase  {
```

Comment:
I think you should just get rid of this templated intermediate class. It only exists to declare these function pointers, but, it never actually calls them (it can't, it doesn't know what the type is). I think it would be better to just call the pipeline definitions directly in the particular Descriptor::run implementation.